File: gpvdm_gui/gui/component.py
# 
#   General-purpose Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
#   model for 1st, 2nd and 3rd generation solar cells.
#   Copyright (C) 2012-2017 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
#
#   https://www.gpvdm.com
#   Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License v2.0, as published by
#   the Free Software Foundation.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License along
#   with this program; if not, write to the Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# 

## @package component
#  An electrical component
#


import os
import math

class component():
	def __init__(self):
		self.name="resistor"
		self.x0=2
		self.y0=2
		self.x1=2
		self.y1=3
		self.z2=2
		self.z2=3

		self.dir="north"
		self.lines=[]

		self.R=10.0
		self.C=0.0
		self.L=0.0
		self.nid=1.0
		self.J0=0.0

	def __str__(self):
		return str(self.x0)+" "+str(self.y0)+" "+str(self.x1)+" "+str(self.y1)+" "+self.name

	def __eq__(self,a):
		if self.x0==a.x0:
			if self.y0==a.y0:
				if self.x1==a.x1:
					if self.y1==a.y1:
						return True

		if self.x0==a.x1:
			if self.y0==a.y1:
				if self.x1==a.x0:
					if self.y1==a.y0:
						return True
		return False

	def get_direction(self):
		if self.x0==self.x1:
			if self.y1>self.y0:
				return "up"
			else:
				return "down"

		if self.y0==self.y1:
			if self.x1>self.x0:
				return "right"
			else:
				return "left"

File: gpvdm_core/libcircuit/component.c
//
// General-purpose Photovoltaic Device Model gpvdm.com - a drift diffusion
// base/Shockley-Read-Hall model for 1st, 2nd and 3rd generation solarcells.
// The model can simulate OLEDs, Perovskite cells, and OFETs.
// 
// Copyright 2008-2022 Roderick C. I. MacKenzie https://www.gpvdm.com
// r.c.i.mackenzie at googlemail.com
// 
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense, 
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
// SOFTWARE.
// 

/** @file component.c
@brief Deals with electrical components
*/

#include "device.h"
#include "util.h"
#include "gpvdm_const.h"
#include <log.h>
#include <cal_path.h>
#include <circuit.h>
#include <matrix.h>
#include <memory.h>
#include <contacts.h>
#include <epitaxy_struct.h>
#include <epitaxy.h>
#include <exp.h>
#include <dump.h>
#include <lang.h>
#include <component.h>
#include <component_fun.h>

void component_init(struct simulation *sim,struct component *com)
{
	strcpy(com->component,"none");
	com->shunt=-1.0;
	com->series_z=-1.0;
	com->series_x=-1.0;
	com->series_y=-1.0;
}

void component_load_electrical_file(struct simulation *sim,struct component *com,struct json_obj *json_obj)
{
	char temp[200];

	if (json_obj==NULL)
	{
		ewe(sim,"Error loading json file for electrical components\n");
		com->shunt=0.0;
		com->series_z=0.0;
		com->series_x=0.0;
		com->series_y=0.0;
		com->n_ideality=0.0;
		com->J0=0.0;
	}else
	{
		json_get_string(sim, json_obj, com->component,"electrical_component");

		json_get_long_double(sim,json_obj, &(com->shunt),"electrical_shunt");

		json_get_string(sim, json_obj, temp,"electrical_symmetrical_resistance");

		json_get_long_double(sim,json_obj, &(com->series_z),"electrical_series_z");

		json_get_long_double(sim,json_obj, &(com->series_x),"electrical_series_x");

		json_get_long_double(sim,json_obj, &(com->series_y),"electrical_series_y");

		if (strcmp(temp,"symmetric")==0)
		{
			com->series_z=com->series_y;
			com->series_x=com->series_y;
		}

		json_get_long_double(sim,json_obj, &(com->n_ideality),"electrical_n");
	
		json_get_long_double(sim,json_obj, &(com->J0),"electrical_J0");

		json_get_english(sim,json_obj, &(com->electrical_enable_generation),"electrical_enable_generation");


	}
}


